homeassistant/components/econet/__init__.py

```diff
@@ -1 +1,194 @@
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+INTERVAL = timedelta(minutes=60)
+
+
+async def async_setup(hass, config):
+    """Set up the EcoNet component."""
+
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][API_CLIENT] = {}
+    hass.data[DOMAIN][EQUIPMENT] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+    _LOGGER.error(conf)
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data={
+                CONF_EMAIL: conf[CONF_EMAIL],
+                CONF_PASSWORD: conf[CONF_PASSWORD],
+            },
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up EcoNet as config entry."""
+    entry_updates = {}
+    if not config_entry.unique_id:
+        # If the config entry doesn't already have a unique ID, set one:
+        entry_updates["unique_id"] = config_entry.data[CONF_EMAIL]
+    if entry_updates:
+        hass.config_entries.async_update_entry(config_entry, **entry_updates)
+
+    email = config_entry.data.get(CONF_EMAIL)
+    password = config_entry.data.get(CONF_PASSWORD)
+
+    try:
+        api = await EcoNetApiInterface.login(email, password=password)
+    except InvalidCredentialsError:
+        _LOGGER.error("Invalid credentials provided")
+        return False
+    except PyeconetError as err:
+        _LOGGER.error("Config entry failed: %s", err)
+        raise ConfigEntryNotReady from err
+
+    equipment = await api.get_equipment_by_type([EquipmentType.WATER_HEATER])
+    hass.data[DOMAIN][API_CLIENT][config_entry.entry_id] = api
+    hass.data[DOMAIN][EQUIPMENT][config_entry.entry_id] = equipment
+
+    for component in ["water_heater"]:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
+        )
+
+    api.subscribe()
+
+    async def resubscribe(now):
+        """Resubscribe to the MQTT updates."""
+        api.unsubscribe()
+        api.subscribe()
+
+    async def fetch_update(now):
+        """Fetch the latest changes from the API."""
+        api.refresh_equipment()
+
+    async_track_time_interval(hass, resubscribe, INTERVAL)
+    async_track_time_interval(hass, fetch_update, INTERVAL + timedelta(minutes=1))
+
+    return True
+
+
+async def async_unload_entry(hass, entry):
+    """Unload a EcoNet config entry."""
+    tasks = [
+        hass.config_entries.async_forward_entry_unload(entry, component)
+        for component in ["water_heater"]
+    ]
+
+    await asyncio.gather(*tasks)
+
+    hass.data[DOMAIN][API_CLIENT].pop(entry.entry_id)
+
+    return True
+
+
+class EcoNetEntity(Entity):
+    """Define a base EcoNet entity."""
+
+    def __init__(self, econet):
+        """Initialize."""
+        self._econet = econet
+        self._econet.set_update_callback(self.on_update_received)
```

Comment:
Register state update callbacks in entity coroutine method `async_added_to_hass`.

Reply:
done